src/stream/src/common/builder.rs

```diff
@@ -133,13 +145,14 @@ impl StreamChunkBuilder {
         row_update: &RowRef<'_>,
     ) -> ArrayResult<Option<StreamChunk>> {
         self.ops.push(op);
-        for (i, d) in row_update.values().enumerate() {
-            self.column_builders[i + self.update_start_pos].append_datum_ref(d)?;
+        for &(update_idx, output_idx) in &self.update_indices_mapping {
+            self.column_builders[output_idx].append_datum_ref(row_update.value_at(update_idx))?;
         }
-        for i in 0..self.column_builders.len() - row_update.size() {
-            self.column_builders[i + self.matched_start_pos].append_datum(&None)?;
+        if !self.is_semi_anti_join {
```

Comment:
For semi and anti join, the `update_range` and the `matched_range` are the same. (Not sure about this, correct me if I'm wrong. I've got this conclusion through the code below) 
```rs
        let (update_start_pos, matched_start_pos) = if is_semi_or_anti(T) {
            (0, 0) // update & match share the same start pos
        } else {
            (side_update.start_pos, side_match.start_pos)
        };
```

This means that the `update_range` and the `matched_range` are overlapped. Therefore the `update_indices_mapping` and the `matched_indices_mapping` are overlapped too. Overlapping is strange here because for normal joins, `update_range` and `matched_range` has no intersections.

So for `append_row_update` and `append_row_matched`, I did a workaround here to ensure "Append a row with coming update/matched value and fill the other side with null". Otherwise, a new row with null value will be appended.

Any suggestion on how to make it better? :heart_eyes:

Reply:
Yes, the original code is technically not correct and depends on an unused code path, so it is a bit hacky. I don't think we want the chunk builder to be `is_semi_or_anti` aware, though. Could we just change it to give the range as 0..0 for the non semi/anti side?